feat(fees): add Puffer Vaults fees adapter

brenopolanski opened the pull request:
## Summary

- Closes: https://github.com/DefiLlama/dimension-adapters/issues/5120

Adds a fees adapter for Puffer Vaults that tracks daily changes in vault assets as a proxy for yields and fees. Estimates protocol revenue based on the documented 10% performance fee structure.

## Sources

- Contract deployment: https://etherscan.io/tx/0x9e1ba7737ed1e35e507cb6e0f6ffb00148357340989401e7e980bc8abad98662
- Fee structure: https://docs.puffer.fi/yield/stakers/withdraw
- Contract repo: https://github.com/PufferFinance/puffer-contracts
- Protocol page: https://defillama.com/protocol/puffer-vaults

llamabutler commented:
The puffer-vaults.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees puffer-vaults.ts
    
    🦙 Running PUFFER-VAULTS.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Mon, 15 Dec 2025 16:02:47 GMT
    End Date:	Tue, 16 Dec 2025 16:02:47 GMT
    ---------------------------------------------------
    
    ETHEREUM 👇
    Backfill start time: 31/1/2024
    Daily fees: 0.00
    Daily revenue: 0.00
    Daily protocol revenue: 0.00
    End timestamp: 1765900966 (2025-12-16T16:02:46.000Z)

treeoflife2 reviewed (changes requested):
.

llamabutler commented:
The puffer-vaults.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees puffer-vaults.ts
    
    🦙 Running PUFFER-VAULTS.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Thu, 18 Dec 2025 10:27:13 GMT
    End Date:	Fri, 19 Dec 2025 10:27:13 GMT
    ---------------------------------------------------
    
    ETHEREUM 👇
    Backfill start time: 31/1/2024
    Daily fees: 2.00
    Daily revenue: 0.00
    Daily protocol revenue: 0.00
    End timestamp: 1766140032 (2025-12-19T10:27:12.000Z)

brenopolanski commented:
Hey @treeoflife2, Fixed!

Changed from tracking `totalAssets` difference (which incorrectly counted deposits as yield) to tracking exchange rate changes via `getRate()`. Now uses the same approach as the TVL adapter - calculates yield from rate increase applied to average supply, which properly isolates actual earnings from deposits/withdrawals. Tracking all 3 vaults with their respective accountants.

> **Ref:** _https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/puffer-vaults/index.js_

llamabutler commented:
The puffer-vaults.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees puffer-vaults.ts
    
    🦙 Running PUFFER-VAULTS.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Mon, 22 Dec 2025 23:14:07 GMT
    End Date:	Tue, 23 Dec 2025 23:14:07 GMT
    ---------------------------------------------------
    
    ETHEREUM 👇
    Backfill start time: 31/1/2024
    Daily fees: 705.00
    Daily revenue: 70.00
    Daily protocol revenue: 70.00
    End timestamp: 1766531646 (2025-12-23T23:14:06.000Z)

brenopolanski commented:
> if that's 825 ETH, then
> 
> ETHEREUM 👇 Backfill start time: 31/1/2024 Daily fees: 825.00 Daily revenue: 82.00 Daily protocol revenue: 82.00 End timestamp: 1730505599 (2024-11-01T23:59:59.000Z)
> 
> this is wrong, as these values are always in the usd

Hey @treeoflife2, I updated the code and the adapter returns raw token amounts, not USD. The CLI just displays USD after downstream price conversion.

Fees are computed as the delta of vault TVL using the same `totalSupply × getRate` logic as the TVL adapter, so units stay consistent. `getRate()` updates are discrete, so daily values can appear lumpy.

> **Ref:** _https://github.com/DefiLlama/DefiLlama-Adapters/blob/main/projects/puffer-vaults/index.js_

Let me know if you do prefer a different approach.

treeoflife2 commented on fees/puffer-vaults.ts:
```diff
@@ -63,24 +61,23 @@ const fetch = async (options: FetchOptions) => {
 
-    if (totalYield > 0n) {
-      dailyFees.add(base, totalYield);
+    const totalRewards = tvlEnd - tvlStart;
```
you are again trying to use the tvl difference as yield

brenopolanski replied:
I will send a fix ASAP

llamabutler commented:
The puffer-vaults.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees puffer-vaults.ts
    
    🦙 Running PUFFER-VAULTS.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Tue, 23 Dec 2025 11:34:30 GMT
    End Date:	Wed, 24 Dec 2025 11:34:30 GMT
    ---------------------------------------------------
    
    ETHEREUM 👇
    Backfill start time: 31/1/2024
    Daily fees: 697.00
    Daily revenue: 70.00
    Daily protocol revenue: 70.00
    End timestamp: 1766576069 (2025-12-24T11:34:29.000Z)

brenopolanski commented:
Hey @treeoflife2, I fixed! Yield is now computed from `getRate deltas × avg supply`, not TVL differences, matching the TVL adapter logic without counting deposits/withdrawals.

llamabutler commented:
The puffer-vaults.ts adapter exports:
        
 
 
    > adapters@1.0.0 test
    > ts-node --transpile-only cli/testAdapter.ts fees puffer-vaults.ts
    
    🦙 Running PUFFER-VAULTS.TS adapter 🦙
    ---------------------------------------------------
    Start Date:	Wed, 24 Dec 2025 09:38:15 GMT
    End Date:	Thu, 25 Dec 2025 09:38:15 GMT
    ---------------------------------------------------
    
    ETHEREUM 👇
    Backfill start time: 31/1/2024
    Daily fees: 0.00
    Daily revenue: 0.00
    Daily protocol revenue: 0.00
    Daily supply side revenue: 0.00
    End timestamp: 1766655494 (2025-12-25T09:38:14.000Z)
